tests/test_models.py

```diff
@@ -407,3 +414,54 @@ def test_migration_rates(self):
         self.assertEqual(model.migration_matrix[1][0], 0.003)
         # check these are the only two nonzero entries in the migration matrix
         self.assertEqual(np.sum(np.array(model.migration_matrix) != 0), 2)
+
+
+class TestPopulationSampling(unittest.TestCase):
+    # Create populations to test on
+    _pop1 = stdpopsim.Population("pop0", "Test pop. 0")
+    _pop2 = stdpopsim.Population("pop1", "Test pop. 1",
+                                 sampling_time=10)
+    _pop3 = stdpopsim.Population("pop2", "Test pop. 2",
+                                 sampling_time=None)
+
+    # Create an empty model to hold populations
+    base_mod = stdpopsim.Model()
+    base_mod.populations = [_pop1, _pop2, _pop3]
+
+    def test_num_sampling_populations(self):
+        self.assertEqual(self.base_mod.num_sampling_populations, 2)
+
+    def test_get_samples(self):
+        test_samples = self.base_mod.get_samples(2, 1)
+        self.assertEqual(len(test_samples), 3)
+        # Check for error when prohibited sampling asked for
+        with self.assertRaises(ValueError):
+            self.base_mod.get_samples(2, 2, 1)
+        # Get the population corresponding to each sample
+        sample_populations = [i.population for i in test_samples]
+        # Check sample populations
+        self.assertEqual(sample_populations, [0, 0, 1])
+        # Test sampling times
+        sample_times = [i.time for i in test_samples]
+        self.assertEqual(sample_times, [0, 0, 10])
+
+    # Test that all sampling populations are specified before non-sampling populations
+    # in the model.populations list
+    def test_population_order(self):
+        for model in stdpopsim.all_models():
+            allow_sample_status = [int(p.allow_samples) for p in model.populations]
+            if sum(allow_sample_status) < len(model.populations):
+                first_non_sampling = allow_sample_status.index(0)
```

Comment:
I'm not following the logic here. What about
```
num_sampling = sum(allow_sample_status)
first_non_sampling = allow_sample_status.index(0)
self.assertEqual(sum(allow_sample_status[:first_non_sampling]), num_sampling)
```

Reply:
This is just checking that the total number of sampling populations is the same as the number of sampling populations up to the index of the first non-sampling population. In other words that all sampling populations are before all non-sampling populations.

Reply:
Mine just gets the index of the last sampling and the first non-sampling populations and checks that the non-sampling comes after the sampling. I'm happy to use your test code instead if you'd prefer.

Reply:
It just wasn't obvious to me what it was doing @ndukler, so it'd be nice to make it more obvious. It's even simpler, right?
```python
allow_sample_status = [int(p.allow_samples) for p in model.populations]
num_sampling = sum(allow_sample_status)
# All sampling populations must be at the start of the list
self.assertEqual(sum(allow_sample_status[num_sampling:]), 0)
```

Reply:
I like this version a lot! I think the previous one may have had an issue when all the samples were sample-able depending on the behavior of `.index(0)` when there was no zero in the vector.